aten/src/ATen/native/ReduceOpsUtils.h

```diff
@@ -148,11 +145,24 @@ static void allocate_reduction_result(
       }
     }
   }
-  if (result.defined()) {
-    result.resize_(shape);
-  } else {
-    result = at::empty(shape, self.options().dtype(dtype));
-  }
+  return shape;
+}
+
+static void allocate_reduction_result(
+    Tensor& result, const Tensor& self, DimMask mask, bool keepdim,
+    ScalarType dtype)
+{
+  auto shape = shape_from_dim_mask(self, mask, keepdim);
+  TORCH_CHECK(result.defined(), "Cannot create a new tensor inside a reduction op. You likely tried to call an operator with an out argument but the out argument was an undefined tensor.");
```

Comment:
`INTERNAL_ASSERT` me

Reply:
This can be triggered by user code without a bug in PyTorch when they call an out overload from the C++ API with an undefined tensor. So this should stay a TORCH_CHECK.